test/parallel/test-tls-add-context.js

```diff
@@ -0,0 +1,74 @@
+'use strict';
+const common = require('../common');
+
+if (!common.hasCrypto)
+  common.skip('missing crypto');
+
+const fixtures = require('../common/fixtures');
+const assert = require('assert');
+const tls = require('tls');
+
+function loadPEM(n) {
+  return fixtures.readKey(`${n}.pem`);
+}
+
+const serverOptions = {
+  key: loadPEM('agent2-key'),
+  cert: loadPEM('agent2-cert'),
+  ca: [ loadPEM('ca2-cert') ],
+  requestCert: true,
+  rejectUnauthorized: false,
+};
+
+const server = tls.createServer(serverOptions, (c) => {
+  if (c.servername === 'unknowncontext') {
+    assert.strictEqual(c.authorized, false);
+    return;
+  }
+  assert.strictEqual(c.authorized, true);
+});
+
+const secureContext = {
+  key: loadPEM('agent1-key'),
+  cert: loadPEM('agent1-cert'),
+  ca: [ loadPEM('ca1-cert') ],
+};
+server.addContext('context1', secureContext);
+server.addContext('context2', tls.createSecureContext(secureContext));
+
+const clientOptionsBase = {
+  key: loadPEM('agent1-key'),
+  cert: loadPEM('agent1-cert'),
+  ca: [ loadPEM('ca1-cert') ],
+  rejectUnauthorized: false,
+};
+
+server.listen(0, common.mustCall(() => {
+  const client1 = tls.connect({
+    ...clientOptionsBase,
+    port: server.address().port,
+    servername: 'context1',
+  }, common.mustCall(() => {
+    client1.end();
+  }));
+
+  const client2 = tls.connect({
+    ...clientOptionsBase,
+    port: server.address().port,
+    servername: 'context2',
+  }, common.mustCall(() => {
+    client2.end();
+  }));
+
+  const client3 = tls.connect({
+    ...clientOptionsBase,
+    port: server.address().port,
+    servername: 'unknowncontext',
+  }, common.mustCall(() => {
+    client3.end();
+  }));
+
+  client3.on('close', common.mustCall(() => {
+    server.close();
```

Comment:
Instead of closing here, can we close the server, server-side when the third connection is made? In this way we avoid a possible race where the server is closed before the first two connections are made.

Reply:
Good point, thanks for bringing it up!

But if we close on the server side when the 3rd connection is made, it also doesn't guarantee the first 2 connections have been connected. Calling `tls.connect()` doesn't mean a connection has been made, CMIIW.

An alternative is:
To await for all clients to be `finished` before calling `server.close()`.
```js
  await Promise.all([
    finished(client1),
    finished(client2),
    finished(client3),
  ]);
  server.close();
```
I think it is also cleaner.

Reply:
> But if we close on the server side when the 3rd connection is made, it also doesn't guarantee the first 2 connections have been connected.

I meant to close when the number of connections is 3, for example:

```js
let connections = 0;

server.on('secureConnection', function () {
  if (++connections === 3) {
    server.close();
  }
});
```

Reply:
Ah, misunderstood it.

Yes, that works, thanks! Rebased & re-tested.